### PR TITLE
Feature/initial kafka matching engine bridging

### PR DESCRIPTION
**Pull request by Ujang360**
@dendisuhubdy, this PR addresses https://github.com/bitwyre/message_queue/issues/1. I will create the persistent message delivery in a separate PR.

Please review after the checklist completed:
- [x] Basic configuration for bridging `REST Server` and `Matching Engine`
- [x] Handle graceful shutdown
- [x] Subscribing to `REST Server`'s Kafka
- [x] Push `FIX Message` received from `REST Server`'s producer to `Matching Engine`
- [x] Travis CI configuration for `NUnit` and `dotnet core`
- [x] Unit Tests

### CODE REVIEW

**Comment by Ujang360** [comment]
@dendisuhubdy I just finished working this feature branch, there are several things that `out-of-scope` for this PR:
1. Test Units for the business logic, because I need to re-design the classes, will do it as soon as we have test this MVP.
2. Persistent message delivery to `Matching Engine`.

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+language: csharp
+mono: none
+dotnet: 2.1.700
+solution: Bitwyre.FIXBridge.sln
+script:
+ - dotnet restore
+ - dotnet test
```
Comment (dendisuhubdy) [inline .travis.yml]: :+1:

File: Bitwyre.FIXBridge.Tests/Bitwyre.FIXBridge.Tests.csproj [hunk 2]
```diff
@@ -0,0 +1,23 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
+
+    <IsPackable>false</IsPackable>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="nunit" Version="3.10.1" />
```
Comment (dendisuhubdy) [inline Bitwyre.FIXBridge.Tests/Bitwyre.FIXBridge.Tests.csproj]: ini unit test packagenya ya?

File: Bitwyre.FIXBridge.Tests/Bitwyre.FIXBridge.Tests.csproj [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
+
+    <IsPackable>false</IsPackable>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="nunit" Version="3.10.1" />
+    <PackageReference Include="NUnit3TestAdapter" Version="3.10.0" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.8.0" />
```
Comment (dendisuhubdy) [inline Bitwyre.FIXBridge.Tests/Bitwyre.FIXBridge.Tests.csproj]: Bedanya yang .NET.Test.Sdk sama dua diatas apa bro?

File: Bitwyre.FIXBridge.Tests/Extensions/StringExtensionsTest.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        [Test]
+        public void StringExtensions_ToIPEndPoint_CorrectlyParseIPEndpoint()
+        {
+            var ipEndpointString0 = "127.0.0.1:8080";
+            var ipEndpointString1 = "localhost:666";
+            var ipEndpointString2 = "0.0.0.0:80";
+            var expectedIPEndpoint0 = new IPEndPoint(IPAddress.Parse("127.0.0.1"), 8080);
+            var expectedIPEndpoint1 = new IPEndPoint(IPAddress.Loopback, 666);
+            var expectedIPEndpoint2 = new IPEndPoint(IPAddress.Any, 80);
+
```
Comment (dendisuhubdy) [inline Bitwyre.FIXBridge.Tests/Extensions/StringExtensionsTest.cs]: parsing IP for endpoint looks great

File: Bitwyre.FIXBridge/App.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+                    "Could not get correct EntryAssembly, please check dotnet core runtime configuration!");
+            }
+
+            _cancellationSource = new CancellationTokenSource();
+            WorkingFolder = Path.GetDirectoryName(entryAssembly.Location);
+            _currentConfig = BridgeConfig.LoadConfig();
+            _messageQueueConsumer = MessageConsumerFactory
+                .CreateMessageConsumer<KafkaConsumer>(new TransparentPushStrategy());
+
+            await Task.WhenAll(_messageQueueConsumer.ConsumeLoopAsync(_currentConfig, _cancellationSource.Token));
```
Comment (dendisuhubdy) [inline Bitwyre.FIXBridge/App.cs]: I see waiting nya pakai `await`?

File: Bitwyre.FIXBridge/App.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        internal static async void ShutdownBridge()
+        {
+            _cancellationSource.Cancel(false);
+
+            if (_messageQueueConsumer != null)
+            {
+                await Task.WhenAll(_messageQueueConsumer.WaitForCompleteShutdown());
```
Comment (dendisuhubdy) [inline Bitwyre.FIXBridge/App.cs]: nah wait, itu jadi kalau messageQueueConsumernya ngga kosong, nunggu status messaged buat nunggu complete shutdown? Itu behaviournya begimane?

File: Bitwyre.FIXBridge/Bitwyre.FIXBridge.csproj [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
+    <LangVersion>7.1</LangVersion>
+    <StartupObject>Bitwyre.FIXBridge.App</StartupObject>
+    <Company>Bitwyre Technologies, LLC.</Company>
+    <Authors>Bitwyre Engineering</Authors>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Confluent.Kafka" Version="1.0.0" />
```
Comment (dendisuhubdy) [inline Bitwyre.FIXBridge/Bitwyre.FIXBridge.csproj]: nice thanks!

File: Bitwyre.FIXBridge/Config/BridgeConfig.cs [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+using Bitwyre.FIXBridge.Extensions;
+using Newtonsoft.Json.Converters;
+
+namespace Bitwyre.FIXBridge.Config
+{
+    public class BridgeConfig
+    {
+        public const string ConfigFilename = "config.json";
+        public const string DefaultProducerEndpoint = "127.0.0.1:9092";
+        public const string DefaultMessageTopic = "fix_message_queue";
```
Comment (dendisuhubdy) [inline Bitwyre.FIXBridge/Config/BridgeConfig.cs]: topicnya udah bener

File: Bitwyre.FIXBridge/Config/BridgeConfig.cs [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        public const string DefaultMatchingEngineEndpoint = "127.0.0.1:5001";
+        public const int DefaultConnectionTimeoutMs = 1000;
```
Comment (dendisuhubdy) [inline Bitwyre.FIXBridge/Config/BridgeConfig.cs]: nah ini langsung lo tembak FIX message nya pakai TCP request kesana kan?

File: Bitwyre.FIXBridge/Factories/MessageConsumerFactory.cs [hunk 10]
```diff
@@ -0,0 +1,13 @@
+﻿using Bitwyre.FIXBridge.Interfaces;
+
+namespace Bitwyre.FIXBridge.Factories
+{
+    public static class MessageConsumerFactory
+    {
+        public static IMessageConsumer CreateMessageConsumer<T>(IMessagePushStrategy pushMessageStrategy)
+            where T : IMessageConsumer, new()
+        {
+            return new T { PushMessageStrategy = pushMessageStrategy };
```
Comment (dendisuhubdy) [inline Bitwyre.FIXBridge/Factories/MessageConsumerFactory.cs]: 👍

File: Bitwyre.FIXBridge/Interfaces/IMessageConsumer.cs [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+using System.Threading.Tasks;
+using Bitwyre.FIXBridge.Config;
+
+namespace Bitwyre.FIXBridge.Interfaces
+{
+    public interface IMessageConsumer
+    {
+        IMessagePushStrategy PushMessageStrategy { get; set; }
+        Task ConsumeLoopAsync(BridgeConfig config, CancellationToken cancellationToken);
+        Task WaitForCompleteShutdown();
```
Comment (dendisuhubdy) [inline Bitwyre.FIXBridge/Interfaces/IMessageConsumer.cs]: ok here it is

File: Bitwyre.FIXBridge/KafkaConsumer.cs [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+                try
+                {
+                    while (!cancellationToken.IsCancellationRequested)
+                    {
+                        try
+                        {
+                            var consumeResult = consumer.Consume(cancellationToken);
+                            var receivedMessage = consumeResult.Message.Value;
+                            Console.WriteLine($"Received message from broker:{Environment.NewLine}{receivedMessage}");
+                            await Task.WhenAll(PushMessageStrategy.PushToMatchingEngineAsync(config, tcpClient,
```
Comment (dendisuhubdy) [inline Bitwyre.FIXBridge/KafkaConsumer.cs]: I see nama functionnya buat nembak ke exchange itu `PushToMatchingEngineAsync`

File: Bitwyre.FIXBridge/KafkaConsumer.cs [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+                    Console.WriteLine($"Unsubscribed from \"{config.MessagesTopic}\" at \"{config.ProducerEndpoint}\"");
+                }
+            }
+
+            _isRunning = false;
+        }
+
+        public async Task WaitForCompleteShutdown()
+        {
+            while (_isRunning)
```
Comment (dendisuhubdy) [inline Bitwyre.FIXBridge/KafkaConsumer.cs]: basically ini putting delay 100 ms sampai ada messages masuk, kalau messages sampai ngga masuk 100 ms?

File: Bitwyre.FIXBridge/TransparentPushStrategy.cs [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+                };
+            }
+
+            if (!matchingEngineClient.Connected)
+            {
+                var serverEndpoint = bridgeConfig.MatchingEngineEndpoint.ToIPEndPoint();
+
+                try
+                {
+                    await matchingEngineClient.ConnectAsync(serverEndpoint.Address, serverEndpoint.Port);
```
Comment (dendisuhubdy) [inline Bitwyre.FIXBridge/TransparentPushStrategy.cs]: dope

**Comment by dendisuhubdy** [comment]
LGTM. Feel free to merge @Ujang360 :)

File: Bitwyre.FIXBridge.Tests/Bitwyre.FIXBridge.Tests.csproj (hunk 2 above)
Comment (Ujang360) [reply Bitwyre.FIXBridge.Tests/Bitwyre.FIXBridge.Tests.csproj]: Yess

File: Bitwyre.FIXBridge.Tests/Bitwyre.FIXBridge.Tests.csproj (hunk 3 above)
Comment (Ujang360) [reply Bitwyre.FIXBridge.Tests/Bitwyre.FIXBridge.Tests.csproj]: Yang atas library untuk test conveniency, yang bawah framework testing nya

File: Bitwyre.FIXBridge.Tests/Extensions/StringExtensionsTest.cs (hunk 4 above)
Comment (Ujang360) [reply Bitwyre.FIXBridge.Tests/Extensions/StringExtensionsTest.cs]: Thank you @dendisuhubdy

File: Bitwyre.FIXBridge/App.cs (hunk 5 above)
Comment (Ujang360) [reply Bitwyre.FIXBridge/App.cs]: `await` di combo sama `WhenAll` itu blocking sebenernya (synchronous), aku bikin `ConsumeLoopAsync` just in-case mau pake pattern `Task-based Asynchronous Pattern` nya Microsoft di kemudian hari.

File: Bitwyre.FIXBridge/App.cs (hunk 6 above)
Comment (Ujang360) [reply Bitwyre.FIXBridge/App.cs]: Nah sebenernya ga ada case _messageQueueConsumer nya kosong, cuman semacam habbit. Basically code ini nunggu consumer selesai aktifitas terakhir sebelum dapet sinyal `Cancel` atau `Quit`. Baru setelah itu dia quit.

File: Bitwyre.FIXBridge/Config/BridgeConfig.cs (hunk 9 above)
Comment (Ujang360) [reply Bitwyre.FIXBridge/Config/BridgeConfig.cs]: Yes mas bro, via `IPushMessageStrategy`.

File: Bitwyre.FIXBridge/KafkaConsumer.cs (hunk 12 above)
Comment (Ujang360) [reply Bitwyre.FIXBridge/KafkaConsumer.cs]: Iya, aku rename. Tadinya `PushToExchangeServerAsync`, salah ga?

File: Bitwyre.FIXBridge/KafkaConsumer.cs (hunk 13 above)
Comment (Ujang360) [reply Bitwyre.FIXBridge/KafkaConsumer.cs]: Nah, basically ini nunggu (blocking) sampe state `_isRunning` nya jadi false. Method ini dipake class lain buat grace shutdown.

**Comment by Ujang360** [comment]
Thank you @dendisuhubdy, lets test this code.

**Pull request merged**